libcodechecker/log/log_parser.py

```diff
@@ -0,0 +1,791 @@
+    """
+    if flag_iterator.item[0] != '-':
+        buildaction.sources.append(flag_iterator.item.strip('"'))
+        return True
+
+    return False
+
+
+def __determine_action_type(flag_iterator, buildaction):
+    """
+    This function determines whether this is a preprocessing, compilation or
+    linking action and sets it in the buildaction object.
+    """
+    if flag_iterator.item == '-c':
+        buildaction.action_type = BuildAction.COMPILE
+        return True
+    elif flag_iterator.item.startswith('-print-prog-name'):
+        buildaction.action_type = BuildAction.INFO
+        return True
+    elif re.match('-(E|M[T|Q|F|J|P|V|M]*)$', flag_iterator.item):
+        buildaction.action_type = BuildAction.PREPROCESS
+        return True
+
+    return False
+
+
+def __get_arch(flag_iterator, buildaction):
+    """
+    This function consumes -arch flag which is followed by the target
+    architecture. This is then collected to the buildaction object.
+    """
+    # TODO: Is this really a target architecture? Have we seen this flag being
+    # used in a real project? This -arch flag is not really documented among
+    # GCC flags.
+    # Where do we use this architecture during analysis and why?
+    if flag_iterator.item == '-arch':
+        next(flag_iterator)
+        buildaction.target = flag_iterator.item
+        return True
+
+    return False
+
+
+def __get_language(flag_iterator, buildaction):
+    """
+    This function consumes -x flag which is followed by the language. This
+    language is then collected to the buildaction object.
+    """
+    # TODO: Known issue: a -x flag may precede all source files in the build
+    # command with different languages.
+    if flag_iterator.item == '-x':
+        next(flag_iterator)
+        buildaction.lang = flag_iterator.item
+        return True
+
+    return False
+
+
+def __get_output(flag_iterator, buildaction):
+    """
+    This function consumes -o flag which is followed by the output file of the
+    action. This file is then collected to the buildaction object.
+    """
+    if flag_iterator.item == '-o':
+        next(flag_iterator)
+        buildaction.output = flag_iterator.item
+        return True
+
+    return False
+
+
+def __replace(flag_iterator, buildaction):
+    """
+    This function extends the analyzer options list with the corresponding
+    replacement based on REPLACE_OPTIONS_MAP if the flag_iterator is currently
+    pointing to a flag to replace.
+    """
+    value = REPLACE_OPTIONS_MAP.get(flag_iterator.item)
+
+    if value:
+        buildaction.analyzer_options.extend(value)
+
+    return bool(value)
+
+
+def __skip(flag_iterator, buildaction):
+    """
+    This function skips the flag pointed by the given flag_iterator with its
+    parameters if any.
+    """
+    if IGNORED_OPTIONS.match(flag_iterator.item):
+        return True
+
+    for pattern, arg_num in IGNORED_PARAM_OPTIONS.items():
+        if re.match(pattern, flag_iterator.item):
```

Comment:
We would compiler the pattern for the regular expression here in the hot loop. Maybe we should first compile each key in the map once when this module is loaded?